hudi-common/src/main/java/org/apache/hudi/common/model/HoodieAvroRecord.java

```diff
@@ -189,14 +191,21 @@ public Option<Map<String, String>> getMetadata() {
 
   @Override
   public Option<HoodieAvroIndexedRecord> toIndexedRecord(Schema recordSchema, Properties props) throws IOException {
-    Option<IndexedRecord> avroData = getData().getInsertValue(recordSchema, props);
+    Option<IndexedRecord> avroData = getDeserData(recordSchema, props);
     if (avroData.isPresent()) {
       return Option.of(new HoodieAvroIndexedRecord(avroData.get()));
     } else {
       return Option.empty();
     }
   }
 
+  private Option<IndexedRecord> getDeserData(Schema recordSchema, Properties props) throws IOException {
+    if (deserRecord == null) {
+      this.deserRecord = this.data.getInsertValue(recordSchema, props);
+    }
+    return this.deserRecord;
```

Comment:
We will also have to do following 
 - Keep the schema this record has been deserialized with, and 
 - Make sure schemas are identical (comparing schemas w/ equals will kill perf-gain so we should compare them as strings)